Laracon 2024

taylorotwell opened the pull request.

taylorotwell committed:
starting concurrency

taylorotwell committed:
wip

taylorotwell committed:
formatting

taylorotwell committed:
driver based

taylorotwell committed:
formatting

taylorotwell committed:
work on defer

taylorotwell committed:
refine defer

taylorotwell committed:
refactor

taylorotwell committed:
formatting

taylorotwell committed:
fix new

taylorotwell committed:
fix status code

taylorotwell committed:
handle job releases after exceptions

taylorotwell committed:
add job attempted event

taylorotwell committed:
formatting

taylorotwell committed:
formatting
"

taylorotwell committed:
first pass at swr

taylorotwell committed:
rename method

taylorotwell committed:
block on web

taylorotwell committed:
fix docblock

taylorotwell committed:
add swr test

taylorotwell committed:
first pass at local temporary files

taylorotwell committed:
remove imports

taylorotwell committed:
docblock

taylorotwell committed:
work on routing

taylorotwell committed:
allow customization of serve behavior

taylorotwell committed:
work on local file serving

taylorotwell committed:
route caching

taylorotwell committed:
move route caching check

taylorotwell committed:
extract file

taylorotwell committed:
formatting

taylorotwell committed:
formatting

taylorotwell committed:
formatting

taylorotwell committed:
add test

taylorotwell committed:
formatting

taylorotwell committed:
add cache control

taylorotwell committed:
use defer in concurrency

taylorotwell committed:
allow return from sleep

taylorotwell committed:
adjust sleep behavior

taylorotwell committed:
add config file

taylorotwell committed:
testing

taylorotwell committed:
add while to sleep

taylorotwell committed:
add autoload

taylorotwell committed:
add function

StyleCIBot committed:
Apply fixes from StyleCI

Jubeki commented on src/Illuminate/Concurrency/composer.json:
```diff
@@ -0,0 +1,35 @@
+        }
+    ],
+    "require": {
+        "php": "^8.0.2",
```
```suggestion
        "php": "^8.2",
```

Illuminate/Process v11 requires PHP 8.2:
https://github.com/illuminate/process/blob/69df5be9de8389a2aed90dc2555b898ee0f3d231/composer.json#L17

Jubeki commented on composer.json:
You should also add `"illuminate/concurrency": "self.version",` to the laravel/framework composer.json

crynobone committed:
Concurrency Improvements (#52713)

* Concurrency Improvements

1. Add a working tests for Concurrency
2. Update composer.json to include replace `illuminate/concurrency`
3. Update PHP versions

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* Apply fixes from StyleCI

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

---------

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>
Co-authored-by: StyleCI Bot <bot@styleci.io>

chu121su12 commented on src/Illuminate/Cache/Repository.php:
```diff
@@ -471,6 +471,61 @@ public function rememberForever($key, Closure $callback)
+     * @param  array{ seconds?: int, owner?: string }|null  $lock
+     * @return TCacheValue
+     */
+    public function flexible($key, $ttl, $callback, $lock = null)
```
Another is why a 2-element array for ttl and not spreading it as, say, `($key, $ttlFresh, $ttlStale, ...)`

devfrey commented on src/Illuminate/Filesystem/FilesystemServiceProvider.php:
```diff
@@ -60,6 +71,49 @@ protected function registerManager()
+                    : '/storage';
+
+                Route::get($uri.'/{path}', function (Request $request, string $path) use ($disk, $config) {
+                    return (new ServeFile(
+                        $disk,
+                        $config,
+                        $this->app->isProduction()
+                    ))($request, $path);
```
Would it be possible to resolve this via the container? That would allow for easier modification.

taylorotwell replied:
I believe you could use `Storage::serveUsing(fn)`

heychazza commented:
Hey @taylorotwell, thanks for such a fantastic Laracon! The `Concurrency` stuff has been a huge feature I'm most excited for, so I gave it a go using your PR to tinker.

I'm not sure if intentional, but the `php` binary at https://github.com/laravel/framework/blob/laracon-2024/src/Illuminate/Concurrency/ProcessDriver.php#L31 is never found when using Laravel Herd.

I presume it's intended that we only use `sync` during local dev, and `process` for production? _I understand this is still a draft, but if release is tomorrow I wanted to point out_.

Thanks a ton for this regardless :)

donnysim commented:
@taylorotwell the PHP binary should definitely be configurable, we have a development server where you switch PHP per project basis and the global PHP is not the one that is used for the project and accessed via `php8.2 artisan` etc.. Maybe introduce a config and env variable?

osbre commented:
I wonder if Octane would have its own concurrency driver since Swoole/OpenSwoole have a feature called "task workers".

taylorotwell committed:
adjust php binary

taylorotwell committed:
Merge branch 'laracon-2024' of github.com:laravel/framework into laracon-2024

taylorotwell committed:
php executable finder

taylorotwell committed:
false arg

devajmeireles commented:
@taylorotwell 

![CleanShot 2024-09-11 at 18 55 45](https://github.com/user-attachments/assets/605edb15-5985-4888-9c0d-f18806b36d84)

Also, you forgot to add the documentation blocks for the Competition facade.

Tklaversma commented on src/Illuminate/Foundation/Defer/DeferredCallbackCollection.php:
```diff
@@ -0,0 +1,157 @@
+    protected function forgetDuplicates(): self
+    {
+        $this->callbacks = collect($this->callbacks)
+            ->reverse()
```
@taylorotwell not sure, but I think we don't need the double reverse(), right?

Mohammed-Abodan-Jabri replied:
Thanks

decadence commented:
I'm trying to call `Concurrency::run` but got:

```
Illuminate\Contracts\Container\BindingResolutionException
Unresolvable dependency resolving [Parameter #0 [ <required> $app ]] in class Illuminate\Support\MultipleInstanceManager
```

Also it seems like `defer` uses `sync` driver.

How to configure this facade?

Jubeki commented:
Please open an issue if you have found a problem. Comments in already merged PRs are mostly not looked after. You can also send in a PR to fix the problem directly.

decadence commented:
I'm not sure if this is a issue. Maybe I'm just doing something wrong.

timacdonald commented:
`defer` has a different feature set than dispatching jobs after the response, such as cancelling on bad status codes or always running.

`defer` is not tied to the queuing system. This means that models are not serialized / re-queried, reducing potential database queries and when captured models are deleted within the request the closure will still run.

If you love job classes and those tradeoffs mean jobs are the best answer for you, go for it.
